emails/views.py

```diff
@@ -118,6 +123,159 @@ def _index_DELETE(request_data, user_profile):
+            }
+        )
+        return HttpResponse(
+            'Received SNS message for unsupported Type: %s' % message_type,
+            status=400
+        )
+
+    json_body = json.loads(request.body)
+    try:
+        verified_json_body = verify_from_sns(json_body)
+    except Exception:
+        logger.error(
+            'SNS message with invalid signature',
+            extra={
+                'SigningCertURL': json_body['SigningCertURL'],
+                'Signature': json_body['Signature'],
+            }
+        )
+        return HttpResponse(
+            'Received SNS message with invalid signature: %s' % message_type,
+            status=401
+        )
+
+    return _sns_inbound_logic(topic_arn, message_type, verified_json_body)
+
+
+def _sns_inbound_logic(topic_arn, message_type, json_body):
+    if message_type == 'SubscriptionConfirmation':
+        logger.info(
+            'SNS SubscriptionConfirmation',
+            extra={'SubscribeURL': json_body['SubscribeURL']}
+        )
+        return HttpResponse('Logged SubscribeURL', status=200)
+    if message_type == 'Notification':
+        logger.info(
+            'SNS Notification',
+            extra={'json_body': json_body},
+        )
+        return _sns_notification(json_body)
+
+
+def _sns_notification(json_body):
+    message_json = json.loads(json_body['Message'])
+    notification_type = message_json['notificationType']
+    if notification_type != 'Received':
+        logger.error(
+            'SNS notification for unsupported type',
+            extra={'notification_type': notification_type},
+        )
+        return HttpResponse(
+            'Received SNS notification for unsupported Type: %s' %
+            notification_type,
+            status=400
+        )
+
+    return _sns_mail(message_json)
+
+
+def _sns_mail(message_json):
+    mail = message_json['mail']
+    to_address = parseaddr(mail['commonHeaders']['to'][0])[1]
+    local_portion = to_address.split('@')[0]
+
+    try:
+        relay_address = RelayAddress.objects.get(address=local_portion)
+        if not relay_address.enabled:
+            relay_address.num_blocked += 1
+            relay_address.save(update_fields=['num_blocked'])
+            return HttpResponse("Address does not exist")
+    except RelayAddress.DoesNotExist:
+        # TODO?: if sha256 of the address is in DeletedAddresses,
+        # create a hard bounce receipt rule
+        logger.error('email_relay', extra={'message_json': message_json})
+        return HttpResponse("Address does not exist", status=404)
+
+    logger.info('email_relay', extra={
+        'fxa_uid': (
+            relay_address.user.socialaccount_set.first().uid
+        ),
+        'relay_address_id': relay_address.id,
+        'relay_address': sha256(local_portion.encode('utf-8')).hexdigest(),
+        'real_address': sha256(
+            relay_address.user.email.encode('utf-8')
+        ).hexdigest(),
```

Comment:
I understand that it's encrypted but is there a need to log the `real_address` then?

Reply:
This is for the fraud & abuse pipeline.